packages/flutter_tools/lib/src/vmservice.dart

```diff
@@ -949,19 +943,31 @@ class VM extends ServiceObjectOwner {
     return invokeRpcRaw('_getVMTimeline', timeout: kLongRequestTimeout);
   }
 
-  Future<void> refreshViews() {
+  Future<void> refreshViews({ bool waitForViews = false }) async {
+    assert(waitForViews != null);
+    assert(loaded);
     if (!isFlutterEngine)
-      return Future<void>.value(null);
-    _viewCache.clear();
-    final Completer<void> completer = Completer<void>();
-    final List<Future<ServiceObject>> futures = <Future<ServiceObject>>[];
-    for (Isolate isolate in isolates.toList()) {
-      futures.add(vmService.vm.invokeRpc<ServiceObject>('_flutter.listViews',
+      return;
+    int failCount = 0;
+    while (true) {
+      _viewCache.clear();
+      final List<Future<void>> futures = <Future<void>>[];
+      for (Isolate isolate in isolates.toList()) {
+        futures.add(vmService.vm.invokeRpc<ServiceObject>(
+          '_flutter.listViews',
           timeout: kLongRequestTimeout,
-          params: <String, dynamic> {'isolateId': isolate.id}));
+          params: <String, dynamic> {'isolateId': isolate.id},
+        ));
+      }
+      await Future.wait(futures);
```

Comment:
Similarly, can we avoid awaiting in this method and instead let caller await if it decided it needed to?

Reply:
The return value of this method will be a future that will complete when all the futures have completed.  It's still up to the caller to await the future returned by this method or not.

Reply:
`await future; ...rest of method...` is exactly equivalent to `return future.then(...rest of method...)`.